### ISSUE TITLE
DBIOManager doesn't work with self-dependent assets

### ISSUE THREAD

**Issue by mjclarke94** [dagster-io/dagster#11845]
### Dagster version

dagster, version 1.1.13

### What's the issue?

When a partitioned asset is first materialised using a DBIOManager, a corresponding table is created in the target database. 

This behaviour breaks down when the asset depends on a past partition of itself. I think the IOManager isn't respecting the `start_time` parameter of the `partition_def`, and so instead of `None` being passed in as an input to the first instance of the asset (as there are no prior partitions), the IOManager tries and fails to access a table which does not yet exist.

### What did you expect to happen?

DBIOManagers should respect the `start_date` parameter of a `TimeWindowPartitionDefinition`, allowing the example given [here](https://gist.github.com/slopp/c4eee86c55c5d5635ccba196672e5ea2) to work with a DBIOManager set.

### How to reproduce?

```
from dagster import (
    asset,
    AssetIn,
    TimeWindowPartitionMapping,
    DailyPartitionsDefinition,
    Definitions,
)
import pandas as pd
from pendulum import datetime
from dagster_duckdb import build_duckdb_io_manager
from dagster_duckdb_pandas import DuckDBPandasTypeHandler

# Create IOManager
duckdb_io_manager = build_duckdb_io_manager([DuckDBPandasTypeHandler()])
resources = {"io_manager": duckdb_io_manager.configured({"database": "test.duckdb"})}

# Create partitions
daily_partitions = DailyPartitionsDefinition(start_date=datetime(2023, 1, 1))


@asset(
    partitions_def=daily_partitions,
    io_manager_key="io_manager",
    ins={
        "broken_asset": AssetIn(
            partition_mapping=TimeWindowPartitionMapping(
                start_offset=-1, end_offset=-1
            ),
        ),
    },
    metadata={
        "partition_expr": "time",
    },
)
def broken_asset(context, broken_asset: pd.DataFrame) -> pd.DataFrame:
    context.log.info("Made it to here.")
    start, end = context.output_asset_partitions_time_window()
    return pd.DataFrame({"time": [start, end], "data": [0.1, 0.2]})


@asset(
    partitions_def=daily_partitions,
    io_manager_key="io_manager",
    metadata={
        "partition_expr": "time",
    },
)
def working_asset(context) -> pd.DataFrame:
    start, end = context.output_asset_partitions_time_window()
    return pd.DataFrame({"time": [start, end], "data": [0.1, 0.2]})


defs = Definitions(assets=[broken_asset, working_asset], resources=resources)

```

The working asset materialises/backfills just fine, whilst we get the following issue with the broken asset:

```
dagster._core.errors.DagsterExecutionLoadInputError: Error occurred while loading input "broken_asset" of step "broken_asset": 
  File "/Users/matt/dev/dagster/self-deps/.venv/lib/python3.10/site-packages/dagster/_core/execution/plan/execute_plan.py", line 265, in dagster_event_sequence_for_step
    for step_event in check.generator(step_events):
  File "/Users/matt/dev/dagster/self-deps/.venv/lib/python3.10/site-packages/dagster/_core/execution/plan/execute_step.py", line 344, in core_dagster_event_sequence_for_step
    for event_or_input_value in ensure_gen(
  File "/Users/matt/dev/dagster/self-deps/.venv/lib/python3.10/site-packages/dagster/_core/execution/plan/inputs.py", line 206, in load_input_object
    yield from _load_input_with_input_manager(loader, load_input_context)
  File "/Users/matt/dev/dagster/self-deps/.venv/lib/python3.10/site-packages/dagster/_core/execution/plan/inputs.py", line 860, in _load_input_with_input_manager
    with op_execution_error_boundary(
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/matt/dev/dagster/self-deps/.venv/lib/python3.10/site-packages/dagster/_core/execution/plan/utils.py", line 85, in op_execution_error_boundary
    raise error_cls(
The above exception was caused by the following exception:

duckdb.CatalogException: Catalog Error: Table with name broken_asset does not exist! Did you mean "working_asset"? LINE 1: SELECT * FROM public.broken_asset
```

The issue can be mitigated by manually creating a table with the correct schema, but then there is no way to distinguish a partition with no data, and a partition outside the scope of the partition definition.

### Deployment type

Local

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by sryza** [comment]
@jamiedemaria is this something you'd be up for taking on? I think the issue here is that `DbIOManager` doesn't know what to do when an input corresponds to zero partitions (which happens for the first partition of an asset that depends on partitions of itself).  So we'd basically need to handle that case and return `None` or an empty dataframe.

**Comment by jamiedemaria** [comment]
fixed in release 0.18.2
